xla/hlo/transforms/collectives/while_loop_all_reduce_code_motion_setup.h

```diff
@@ -0,0 +1,60 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/op_expander_pass.h"
+
+namespace xla {
+
+// Reorder the sequence of reduce-scatter, convert, transpose, and add
+// operations. This transformation changes the pattern from:
```

Comment:
Can you comment on why this is beneficial?

Reply:
See comment above - if you reorder the sequence of ops and then WhileLoopAllReduceCodeMotion now I can code motion the all-reduce and reduce-scatter out of the while loop. 
Reduce-scatter is controlled obviously by the existing XLA debug flag, but all-reduce is enabled by default in GPU pipeline, so what this means is 

Say I have a while loop trip count of 4 running 5 all-reduces per while loop iteration, but those all-reduces have a convert or transpose user before the add accumulation op, I can run these new passes and code motion them. Now I saved 20 all-reduces in my train graph.

Reply:
Makes sense. Ty